File: parsers/metadata_parser.py
import json
import re


class MetadataParser:
    def __init__(self):
        self.metadata = {
            'prompt': '',
            'negative_prompt': '',
            'seed': '',
            'steps': '',
            'cfg': '',
            'sampler_name': '',
            'model': '',
            'workflow': ''
        }
    
    def parse(self, img):
        # Check for chara data first (AI Chat Character)
        chara_data = None
        chara_raw = None
        if 'chara' in img.info:
            chara_data = img.info['chara']
            import base64
            try:
                decoded = base64.b64decode(chara_data)
                chara_raw = json.loads(decoded.decode('utf-8'))
            except:
                pass
        
        # Then check for other metadata
        if 'parameters' in img.info:
            result = self.parse_parameters(img.info['parameters'])
        elif 'prompt' in img.info:
            result = self.parse_prompt_json(img.info['prompt'])
        elif 'exif_b64' in img.info:
            # Handle base64 encoded exif data (from PNG saved by our app)
            import base64
            try:
                exif_data = base64.b64decode(img.info['exif_b64'])
                result = self.parse_exif_data(exif_data)
            except:
                result = self.parse_other_ai_image()
        elif 'exif' in img.info:
            result = self.parse_exif_data(img.info['exif'])
        elif 'AIGC' in img.info:
            result = self.parse_aigc_data(img.info['AIGC'])
        elif 'xmp' in img.info:
            result = self.parse_xmp_data(img.info['xmp'])
        elif chara_data:
            result = self.parse_chat_character(chara_data)
        else:
            result = self.parse_other_ai_image()
        
        # If chara data exists, add it to the result
        if chara_data:
            if result.get('model') != 'AI Chat Character':
                chara_result = self.parse_chat_character(chara_data)
                result['chara_workflow'] = chara_result.get('workflow', '')
            # Always save the raw chara data for editing
            if chara_raw:
                result['chara_raw'] = chara_raw
        
        return result
    
    def parse_parameters(self, parameters):
        lines = parameters.split('\n')
        self.metadata = {
            'prompt': '',
            'negative_prompt': '',
            'seed': '',
            'steps': '',
            'cfg': '',
            'sampler_name': '',
            'model': '',
            'workflow': ''
        }

        prompt_lines = []
        in_workflow = False
        workflow_lines = []

        for line in lines:
            line = line.strip()
            if not in_workflow:
                if line.startswith('Negative prompt:'):
                    self.metadata['negative_prompt'] = line.replace('Negative prompt:', '').strip()
                elif line.startswith('Steps:'):
                    parts = line.split(',')
                    for part in parts:
                        part = part.strip()
                        if part.startswith('Steps:'):
                            self.metadata['steps'] = part.replace('Steps:', '').strip()
                        elif part.startswith('CFG scale:'):
                            self.metadata['cfg'] = part.replace('CFG scale:', '').strip()
                        elif part.startswith('Sampler:'):
                            self.metadata['sampler_name'] = part.replace('Sampler:', '').strip()
                        elif part.startswith('Seed:'):
                            self.metadata['seed'] = part.replace('Seed:', '').strip()
                        elif part.startswith('Model:'):
                            self.metadata['model'] = part.replace('Model:', '').strip()
                elif line.startswith('Workflow:'):
                    in_workflow = True
                else:
                    prompt_lines.append(line)
            else:
                workflow_lines.append(line)

        self.metadata['prompt'] = '\n'.join(prompt_lines).strip()
        if workflow_lines:
            self.metadata['workflow'] = '\n'.join(workflow_lines).strip()

        return self.metadata

    def parse_prompt_json(self, prompt_data):
        self.metadata = {
            'prompt': '',
            'negative_prompt': '',
            'seed': '',
            'steps': '',
            'cfg': '',
            'sampler_name': '',
            'model': '',
            'workflow': prompt_data
        }

        try:
            workflow_data = json.loads(prompt_data)
            prompt_parts = []
            negative_parts = []
            
            for node_id, node_data in workflow_data.items():
                class_type = node_data.get('class_type', '')
                inputs = node_data.get('inputs', {})
                title = node_data.get('_meta', {}).get('title', '').lower()
                
                if class_type == 'Text Multiline':
                    text = inputs.get('text', '')
                    if text:
                        if isinstance(text, list):
                            text = ' '.join(map(str, text))
                        prompt_parts.append(text)
                
                elif class_type == 'CLIPTextEncode':
                    text = inputs.get('text', '')
                    if text:
                        if isinstance(text, list):
                            text = ' '.join(map(str, text))
                        if 'negative' in title:
                            negative_parts.append(text)
                        elif 'positive' in title or not negative_parts:
                            prompt_parts.append(text)
                
                elif class_type in ('CLIPTextEncodeSDXL', 'CLIPTextEncodeSDXLRefiner'):
                    text_g = inputs.get('text_g', '')
                    text_l = inputs.get('text_l', '')
                    if text_g:
                        if isinstance(text_g, list):
                            text_g = ' '.join(map(str, text_g))
                        prompt_parts.append(text_g)
                    if text_l:
                        if isinstance(text_l, list):
                            text_l = ' '.join(map(str, text_l))
                        prompt_parts.append(text_l)

            if prompt_parts:
                self.metadata['prompt'] = '\n'.join(prompt_parts)
            if negative_parts:
                self.metadata['negative_prompt'] = '\n'.join(negative_parts)

            for node_id, node_data in workflow_data.items():
                class_type = node_data.get('class_type', '')
                inputs = node_data.get('inputs', {})
                
                if class_type in ('KSampler', 'KSamplerAdvanced'):
                    self.metadata['seed'] = str(inputs.get('seed', inputs.get('noise_seed', '')))
                    self.metadata['steps'] = str(inputs.get('steps', ''))
                    self.metadata['cfg'] = str(inputs.get('cfg', ''))
                    self.metadata['sampler_name'] = str(inputs.get('sampler_name', ''))
                elif class_type in ('UNETLoader', 'CheckpointLoaderSimple'):
                    model_name = inputs.get('unet_name', inputs.get('ckpt_name', ''))
                    if model_name:
                        if isinstance(model_name, list):
                            model_name = str(model_name)
                        self.metadata['model'] = model_name.split('\\')[-1]
        except json.JSONDecodeError:
            pass

        return self.metadata

    def parse_exif_data(self, exif_data):
        self.metadata = {
            'prompt': '',
            'negative_prompt': '',
            'seed': '',
            'steps': '',
            'cfg': '',
            'sampler_name': '',
            'model': '',
            'workflow': ''
        }
        
        try:
            unicode_index = exif_data.find(b'UNICODE')
            if unicode_index == -1:
                encodings = ['utf-8', 'utf-16', 'utf-16-be', 'utf-16-le', 'latin-1']
                best_exif_str = ''
                for encoding in encodings:
                    try:
                        exif_str = exif_data.decode(encoding, errors='ignore')
                        if exif_str and len(exif_str) > len(best_exif_str):
                            best_exif_str = exif_str
                    except:
                        continue
                
                exif_str = best_exif_str
                
                if exif_str:
                    exif_str = ''.join(c for c in exif_str if c.isprintable() or c in '\n\t')
                    self.metadata['workflow'] = f"原始EXIF数据: {exif_str[:500]}"
                    
                    prompt_match = re.search(r'prompt[:=]\s*(.+)', exif_str, re.IGNORECASE)
                    if prompt_match:
                        self.metadata['prompt'] = prompt_match.group(1).strip()
                    else:
                        lines = exif_str.split('\n')
                        long_lines = [line.strip() for line in lines if len(line.strip()) > 20]
                        if long_lines:
                            self.metadata['prompt'] = max(long_lines, key=len)
                        else:
                            prompt_lines = []
                            for line in lines:
                                line = line.strip()
                                if line and not any(keyword in line for keyword in ['Steps:', 'CFG scale:', 'Seed:', 'Sampler:', 'Model:', 'Negative']):
                                    prompt_lines.append(line)
                            if prompt_lines:
                                self.metadata['prompt'] = ' '.join(prompt_lines)
                            else:
                                self.metadata['prompt'] = exif_str[:500]
                    
                    negative_match = re.search(r'negative[:=]\s*(.+)', exif_str, re.IGNORECASE)
                    if negative_match:
                        self.metadata['negative_prompt'] = negative_match.group(1).strip()
                    else:
                        negative_match = re.search(r'Negative\s+prompt[:=]\s*(.+)', exif_str, re.IGNORECASE)
                        if negative_match:
                            self.metadata['negative_prompt'] = negative_match.group(1).strip()
                    
                    if 'Steps:' in exif_str:
                        match = re.search(r'Steps:\s*(\d+)', exif_str)
                        self.metadata['steps'] = match.group(1) if match else ''
                    if 'CFG scale:' in exif_str:
                        match = re.search(r'CFG scale:\s*([\d.]+)', exif_str)
                        self.metadata['cfg'] = match.group(1) if match else ''
                    if 'Seed:' in exif_str:
                        match = re.search(r'Seed:\s*(\d+)', exif_str)
                        self.metadata['seed'] = match.group(1) if match else ''
                    if 'Sampler:' in exif_str:
                        match = re.search(r'Sampler:\s*([^,]+)', exif_str)
                        self.metadata['sampler_name'] = match.group(1) if match else ''
                    if 'Model:' in exif_str:
                        match = re.search(r'Model:\s*([^,]+)', exif_str)
                        self.metadata['model'] = match.group(1) if match else ''
                else:
                    self.metadata['prompt'] = 'Unable to parse prompt'
                    self.metadata['workflow'] = 'Unable to parse EXIF data'
            else:
                unicode_data = exif_data[unicode_index + 8:]
                decoded = None
                
                def count_ascii(s):
                    return sum(1 for c in s if ord(c) < 128)
                
                best_decoded = None
                best_ascii_count = 0
                
                for encoding in ['utf-16-be', 'utf-16-le', 'utf-16']:
                    try:
                        test_decoded = unicode_data.decode(encoding)
                        ascii_count = count_ascii(test_decoded)
                        if ascii_count > best_ascii_count:
                            best_ascii_count = ascii_count
                            best_decoded = test_decoded
                    except:
                        continue
                
                if best_decoded:
                    decoded = best_decoded
                else:
                    decoded = unicode_data.decode('utf-16', errors='ignore')
                
                exif_str = ''.join(c for c in decoded if c.isprintable() or c in '\n\t')
                lines = exif_str.split('\n')
                
                prompt_lines = []
                negative_prompt_lines = []
                params_lines = []
                current_section = 'prompt'
                
                for line in lines:
                    line = line.strip()
                    if not line:
                        continue
                    
                    if 'Negative' in line:
                        current_section = 'negative_prompt'
                        negative_prompt = line.split('Negative', 1)[1].strip()
                        if negative_prompt:
                            negative_prompt_lines.append(negative_prompt)
                    elif line.startswith('Steps:'):
                        current_section = 'params'
                        params_lines.append(line)
                    elif current_section == 'prompt':
                        prompt_lines.append(line)
                    elif current_section == 'negative_prompt':
                        negative_prompt_lines.append(line)
                    elif current_section == 'params':
                        params_lines.append(line)
                
                self.metadata['prompt'] = ' '.join(prompt_lines).strip()
                self.metadata['negative_prompt'] = ' '.join(negative_prompt_lines).strip()
                
                if params_lines:
                    params_text = ' '.join(params_lines)
                    self.metadata['workflow'] = params_text
                    
                    if 'Steps:' in params_text:
                        self.metadata['steps'] = params_text.split('Steps:')[1].split(',')[0].strip()
                    if 'CFG scale:' in params_text:
                        self.metadata['cfg'] = params_text.split('CFG scale:')[1].split(',')[0].strip()
                    if 'Seed:' in params_text:
                        self.metadata['seed'] = params_text.split('Seed:')[1].split(',')[0].strip()
                    if 'Sampler:' in params_text:
                        self.metadata['sampler_name'] = params_text.split('Sampler:')[1].split(',')[0].strip()
                    if 'Model:' in params_text:
                        self.metadata['model'] = params_text.split('Model:')[1].split(',')[0].strip()
            
            if not self.metadata['prompt'] or len(self.metadata['prompt']) < 5:
                if self.metadata['workflow']:
                    workflow_str = self.metadata['workflow']
                    lines = workflow_str.split('\n')
                    long_lines = [line.strip() for line in lines if len(line.strip()) > 20]
                    if long_lines:
                        self.metadata['prompt'] = max(long_lines, key=len)
                    else:
                        non_param_lines = [line.strip() for line in lines if not any(keyword in line for keyword in ['Steps:', 'CFG scale:', 'Seed:', 'Sampler:', 'Model:'])]
                        if non_param_lines:
                            self.metadata['prompt'] = ' '.join(non_param_lines)
                        else:
                            self.metadata['prompt'] = 'Unable to parse prompt'
        except Exception as e:
            self.metadata = {
                'prompt': 'Unable to parse prompt',
                'negative_prompt': '',
                'seed': '',
                'steps': '',
                'cfg': '',
                'sampler_name': '',
                'model': '',
                'workflow': f'Error parsing EXIF data: {str(e)}'
            }

        return self.metadata

    def parse_aigc_data(self, aigc_data):
        self.metadata = {
            'prompt': 'None',
            'negative_prompt': 'None',
            'seed': 'None',
            'steps': 'None',
            'cfg': 'None',
            'sampler_name': 'None',
            'model': 'Doubao',
            'workflow': ''
        }
        
        try:
            aigc_info = json.loads(aigc_data)
            self.metadata['workflow'] = "Doubao AI Info:\n"
            for key, value in aigc_info.items():
                self.metadata['workflow'] += f"{key}: {value}\n"
        except Exception:
            pass

        return self.metadata

    def parse_xmp_data(self, xmp_data):
        self.metadata = {
            'prompt': 'None',
            'negative_prompt': 'None',
            'seed': 'None',
            'steps': 'None',
            'cfg': 'None',
            'sampler_name': 'None',
            'model': 'Qwen Image',
            'workflow': ''
        }
        
        try:
            xmp_str = xmp_data.decode('utf-8', errors='ignore')
            aigc_match = re.search(r'TC260:AIGC="(.*?)"', xmp_str)
            if aigc_match:
                aigc_json = aigc_match.group(1)
                aigc_json = aigc_json.replace('&quot;', '"')
                try:
                    aigc_info = json.loads(aigc_json)
                    self.metadata['workflow'] = "Qwen Image AI Info:\n"
                    for key, value in aigc_info.items():
                        self.metadata['workflow'] += f"{key}: {value}\n"
                except:
                    self.metadata['workflow'] = f"XMP Data: {xmp_str[:500]}..."
            else:
                self.metadata['workflow'] = f"XMP Data: {xmp_str[:500]}..."
        except Exception:
            pass

        return self.metadata

    def parse_chat_character(self, chara_data):
        self.metadata = {
            'prompt': 'AI Chat Character',
            'negative_prompt': 'None',
            'seed': 'None',
            'steps': 'None',
            'cfg': 'None',
            'sampler_name': 'None',
            'model': 'AI Chat Character',
            'workflow': ''
        }
        
        try:
            # Try to decode base64 if it looks like base64
            import base64
            
            # Check if it's base64 encoded
            if chara_data and chara_data[0] in 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=':
                try:
                    decoded = base64.b64decode(chara_data)
                    chara_data = decoded.decode('utf-8')
                except:
                    pass
            
            chara_info = json.loads(chara_data)
            self.metadata['workflow'] = "AI Chat Character Info:\n"
            self.metadata['workflow'] += f"Name: {chara_info.get('name', 'Unknown')}\n"
            self.metadata['workflow'] += f"Description: {chara_info.get('description', '')}\n"
            
            first_mes = chara_info.get('first_mes', '')
            if first_mes:
                self.metadata['workflow'] += f"First Message: {first_mes}\n"
            
            scenario = chara_info.get('scenario', '')
            if scenario:
                self.metadata['workflow'] += f"Scenario: {scenario}\n"
            
            create_date = chara_info.get('create_date', '')
            if create_date:
                self.metadata['workflow'] += f"Create Date: {create_date}\n"
            
            personality = chara_info.get('personality', '')
            if personality:
                self.metadata['workflow'] += f"Personality: {personality}\n"
        except Exception as e:
            self.metadata['workflow'] = f"Error parsing chat character: {str(e)}"
        
        return self.metadata

    def parse_other_ai_image(self):
        self.metadata = {
            'prompt': 'None',
            'negative_prompt': 'None',
            'seed': 'None',
            'steps': 'None',
            'cfg': 'None',
            'sampler_name': 'None',
            'model': 'Unknown AI Tool',
            'workflow': 'No metadata found'
        }
        return self.metadata

File: .web/src/utils/metadataParser.js
export const parseImageMetadata = (file) => {
  return new Promise((resolve) => {
    const reader = new FileReader();
    
    reader.onload = function(event) {
      const arrayBuffer = event.target.result;
      
      try {
        // 尝试从图片中提取真实的元数据
        const metadata = extractRealMetadata(arrayBuffer, file.name);
        resolve(metadata);
      } catch (error) {
        console.error('Error parsing metadata:', error);
        // 解析失败时返回默认值
        resolve({
          prompt: '',
          negative_prompt: '',
          seed: '',
          steps: '',
          cfg: '',
          sampler_name: '',
          model: '',
          workflow: 'Error parsing metadata',
          chara_raw: null
        });
      }
    };
    
    reader.onerror = function() {
      resolve({
        prompt: '',
        negative_prompt: '',
        seed: '',
        steps: '',
        cfg: '',
        sampler_name: '',
        model: '',
        workflow: 'Error reading file',
        chara_raw: null
      });
    };
    
    reader.readAsArrayBuffer(file);
  });
}

function extractRealMetadata(arrayBuffer, filename) {
  const uint8Array = new Uint8Array(arrayBuffer);
  
  // 初始化结果对象（包含chara_raw字段）
  let result = {
    prompt: '',
    negative_prompt: '',
    seed: '',
    steps: '',
    cfg: '',
    sampler_name: '',
    model: '',
    workflow: '',
    chara_raw: null
  };
  
  // 检查是否是WebP格式
  if (uint8Array[0] === 0x52 && uint8Array[1] === 0x49 && 
      uint8Array[2] === 0x46 && uint8Array[3] === 0x46) {
    Object.assign(result, parseWebPMetadata(uint8Array));
  }
  
  // 检查是否是PNG格式
  else if (uint8Array[0] === 0x89 && uint8Array[1] === 0x50 && 
           uint8Array[2] === 0x4E && uint8Array[3] === 0x47) {
    Object.assign(result, parsePNGMetadata(uint8Array));
  }
  
  // 检查是否是JPEG格式
  else if (uint8Array[0] === 0xFF && uint8Array[1] === 0xD8) {
    Object.assign(result, parseJPEGMetadata(uint8Array));
  }
  
  // 尝试提取角色数据（适用于所有格式）
  result.chara_raw = extractCharacterData(uint8Array);
  
  return result;
}

function parseWebPMetadata(uint8Array) {
  let exifData = null;
  
  // WebP格式：查找EXIF chunk
  for (let i = 12; i < uint8Array.length - 4; i++) {
    // 查找 "EXIF" 标记
    if (uint8Array[i] === 0x45 && uint8Array[i+1] === 0x58 &&
        uint8Array[i+2] === 0x49 && uint8Array[i+3] === 0x46) {
      const chunkSize = (uint8Array[i+7] << 24) | (uint8Array[i+6] << 16) | 
                        (uint8Array[i+5] << 8) | uint8Array[i+4];
      exifData = uint8Array.slice(i + 8, i + 8 + chunkSize);
      break;
    }
  }
  
  if (!exifData) {
    return {
      prompt: '',
      negative_prompt: '',
      seed: '',
      steps: '',
      cfg: '',
      sampler_name: '',
      model: '',
      workflow: 'No EXIF data found in WebP'
    };
  }
  
  return parseEXIFData(exifData);
}

function parsePNGMetadata(uint8Array) {
  let textChunks = [];
  let exifB64Data = null;
  let i = 8; // 跳过PNG签名
  
  while (i < uint8Array.length) {
    const length = (uint8Array[i] << 24) | (uint8Array[i+1] << 16) | 
                   (uint8Array[i+2] << 8) | uint8Array[i+3];
    
    if (i + 12 + length > uint8Array.length) break;
    
    const type = String.fromCharCode(...uint8Array.slice(i+4, i+8));
    
    if (type === 'tEXt' || type === 'iTXt' || type === 'zTXt') {
      const data = uint8Array.slice(i + 8, i + 8 + length);
      textChunks.push({ type, data });
      
      // 检查是否是exif_b64字段
      const text = decodeTextChunk(data);
      const nullIndex = text.indexOf('\0');
      if (nullIndex > 0) {
        const keyword = text.substring(0, nullIndex).toLowerCase();
        const value = text.substring(nullIndex + 1);
        
        if (keyword === 'exif_b64') {
          exifB64Data = value;
        }
      }
    }
    
    if (type === 'IEND') break;
    i += 12 + length; // 4(length) + 4(type) + length + 4(CRC)
  }
  
  // 优先处理exif_b64字段（base64编码的EXIF数据）
  if (exifB64Data) {
    try {
      // 解码base64
      const binaryString = atob(exifB64Data);
      const bytes = new Uint8Array(binaryString.length);
      for (let j = 0; j < binaryString.length; j++) {
        bytes[j] = binaryString.charCodeAt(j);
      }
      
      // 解析EXIF数据
      return parseEXIFData(bytes);
    } catch (error) {
      console.error('Error decoding exif_b64:', error);
    }
  }
  
  // 查找parameters或prompt字段
  for (const chunk of textChunks) {
    const text = decodeTextChunk(chunk.data);
    if (chunk.type === 'tEXt' || chunk.type === 'iTXt') {
      const nullIndex = text.indexOf('\0');
      if (nullIndex > 0) {
        const keyword = text.substring(0, nullIndex).toLowerCase();
        const value = text.substring(nullIndex + 1);
        
        if (keyword === 'parameters') {
          return parseParametersText(value);
        }
        if (keyword === 'prompt') {
          // 检查是否是ComfyUI Workflow JSON格式
          try {
            const workflowData = JSON.parse(value);
            return parseComfyUIWorkflow(workflowData);
          } catch (e) {
            // 如果不是JSON，直接返回文本
            return { prompt: value, negative_prompt: '', seed: '', steps: '', cfg: '', sampler_name: '', model: '', workflow: value };
          }
        }
      }
    }
  }
  
  return {
    prompt: '',
    negative_prompt: '',
    seed: '',
    steps: '',
    cfg: '',
    sampler_name: '',
    model: '',
    workflow: 'No metadata found in PNG'
  };
}

function parseJPEGMetadata(uint8Array) {
  let exifOffset = -1;
  
  // 查找APP1标记（EXIF）
  for (let i = 0; i < uint8Array.length - 1; i++) {
    if (uint8Array[i] === 0xFF && uint8Array[i+1] === 0xE1) {
      const segmentLength = (uint8Array[i+2] << 8) | uint8Array[i+3];
      exifOffset = i + 4;
      break;
    }
  }
  
  if (exifOffset === -1) {
    return {
      prompt: '',
      negative_prompt: '',
      seed: '',
      steps: '',
      cfg: '',
      sampler_name: '',
      model: '',
      workflow: 'No EXIF data found in JPEG'
    };
  }
  
  // 跳过 "Exif\0\0" 头部
  const exifData = uint8Array.slice(exifOffset + 6);
  return parseEXIFData(exifData);
}

function parseEXIFData(exifData) {
  try {
    // 将字节数组转换为字符串
    let exifString = '';
    const encodings = ['utf-8', 'utf-16be', 'utf-16le', 'latin1'];
    
    for (const encoding of encodings) {
      try {
        const decoder = new TextDecoder(encoding);
        exifString = decoder.decode(exifData);
        
        // 过滤非打印字符
        exifString = exifString.replace(/[^\x20-\x7E\n\t]/g, '');
        
        if (exifString.length > 10 && /[a-zA-Z]/.test(exifString)) {
          break;
        }
      } catch (e) {
        continue;
      }
    }
    
    if (!exifString || exifString.length < 5) {
      return {
        prompt: '',
        negative_prompt: '',
        seed: '',
        steps: '',
        cfg: '',
        sampler_name: '',
        model: '',
        workflow: 'Unable to decode EXIF data'
      };
    }
    
    // 检查是否有UNICODE标记
    const unicodeIndex = exifString.indexOf('UNICODE');
    if (unicodeIndex !== -1) {
      exifString = exifString.substring(unicodeIndex + 7);
    }
    
    // 解析参数文本格式
    return parseParametersText(exifString);
  } catch (error) {
    console.error('Error parsing EXIF:', error);
    return {
      prompt: '',
      negative_prompt: '',
      seed: '',
      steps: '',
      cfg: '',
      sampler_name: '',
      model: '',
      workflow: 'Error parsing EXIF: ' + error.message
    };
  }
}

function parseParametersText(text) {
  const lines = text.split('\n').map(line => line.trim()).filter(line => line.length > 0);
  
  let promptLines = [];
  let negativePrompt = '';
  let seed = '';
  let steps = '';
  let cfg = '';
  let samplerName = '';
  let model = '';
  let currentSection = 'prompt';
  
  for (const line of lines) {
    if (line.startsWith('Negative prompt:')) {
      currentSection = 'negative';
      negativePrompt = line.substring('Negative prompt:'.length).trim();
    } else if (line.startsWith('Steps:')) {
      currentSection = 'params';
      const parts = line.split(',');
      for (const part of parts) {
        const trimmed = part.trim();
        if (trimmed.startsWith('Steps:')) steps = trimmed.substring('Steps:'.length).trim();
        else if (trimmed.startsWith('CFG scale:')) cfg = trimmed.substring('CFG scale:'.length).trim();
        else if (trimmed.startsWith('Sampler:')) samplerName = trimmed.substring('Sampler:'.length).trim();
        else if (trimmed.startsWith('Seed:')) seed = trimmed.substring('Seed:'.length).trim();
        else if (trimmed.startsWith('Model:')) model = trimmed.substring('Model:'.length).trim();
      }
    } else if (line.startsWith('Workflow:')) {
      currentSection = 'workflow';
    } else if (currentSection === 'prompt') {
      promptLines.push(line);
    }
  }
  
  const prompt = promptLines.join('\n').trim();
  const workflow = lines.join('\n');
  
  return {
    prompt,
    negative_prompt: negativePrompt,
    seed,
    steps,
    cfg,
    sampler_name: samplerName,
    model,
    workflow
  };
}

function decodeTextChunk(data) {
  try {
    // 尝试UTF-8解码
    const decoder = new TextDecoder('utf-8');
    return decoder.decode(data);
  } catch (e) {
    try {
      // 回退到Latin-1
      const decoder = new TextDecoder('latin-1');
      return decoder.decode(data);
    } catch (e2) {
      return '';
    }
  }
}

function parseComfyUIWorkflow(workflowData) {
  let promptParts = [];
  let negativeParts = [];
  let seed = '';
  let steps = '';
  let cfg = '';
  let samplerName = '';
  let model = '';
  
  // 遍历所有节点
  for (const nodeId in workflowData) {
    const node = workflowData[nodeId];
    const classType = node.class_type;
    const inputs = node.inputs || {};
    const meta = node._meta || {};
    const title = (meta.title || '').toLowerCase();
    
    // 提取文本节点
    if (classType === 'Text Multiline') {
      let text = inputs.text || '';
      if (Array.isArray(text)) text = text.join(' ');
      if (text) promptParts.push(text);
    }
    
    // 提取CLIP文本编码节点
    else if (classType === 'CLIPTextEncode') {
      let text = inputs.text || '';
      if (Array.isArray(text)) text = text.join(' ');
      if (text) {
        if (title.includes('negative')) {
          negativeParts.push(text);
        } else {
          promptParts.push(text);
        }
      }
    }
    
    // 提取SDXL文本编码节点
    else if (classType === 'CLIPTextEncodeSDXL' || classType === 'CLIPTextEncodeSDXLRefiner') {
      let textG = inputs.text_g || '';
      let textL = inputs.text_l || '';
      
      if (Array.isArray(textG)) textG = textG.join(' ');
      if (Array.isArray(textL)) textL = textL.join(' ');
      
      if (textG) promptParts.push(textG);
      if (textL) promptParts.push(textL);
    }
    
    // 提取采样器参数
    else if (classType === 'KSampler' || classType === 'KSamplerAdvanced') {
      seed = String(inputs.seed || inputs.noise_seed || '');
      steps = String(inputs.steps || '');
      cfg = String(inputs.cfg || '');
      samplerName = String(inputs.sampler_name || '');
    }
    
    // 提取模型信息
    else if (classType === 'UNETLoader' || classType === 'CheckpointLoaderSimple') {
      let modelName = inputs.unet_name || inputs.ckpt_name || '';
      if (modelName) {
        if (Array.isArray(modelName)) modelName = modelName.toString();
        // 提取文件名（去掉路径）
        model = modelName.split('\\').pop().split('/').pop();
      }
    }
  }
  
  const prompt = promptParts.join('\n').trim();
  const negativePrompt = negativeParts.join('\n').trim();
  
  return {
    prompt,
    negative_prompt: negativePrompt,
    seed,
    steps,
    cfg,
    sampler_name: samplerName,
    model,
    workflow: JSON.stringify(workflowData, null, 2)
  };
}

function extractCharacterData(uint8Array) {
  // 检查是否是PNG格式
  if (uint8Array[0] === 0x89 && uint8Array[1] === 0x50 && 
      uint8Array[2] === 0x4E && uint8Array[3] === 0x47) {
    return extractCharacterFromPNG(uint8Array);
  }
  
  // 其他格式暂不支持角色数据提取
  return null;
}

function extractCharacterFromPNG(uint8Array) {
  let i = 8; // 跳过PNG签名
  
  while (i < uint8Array.length) {
    const length = (uint8Array[i] << 24) | (uint8Array[i+1] << 16) | 
                   (uint8Array[i+2] << 8) | uint8Array[i+3];
    
    if (i + 12 + length > uint8Array.length) break;
    
    const type = String.fromCharCode(...uint8Array.slice(i+4, i+8));
    
    if (type === 'tEXt' || type === 'iTXt') {
      const data = uint8Array.slice(i + 8, i + 8 + length);
      const text = decodeTextChunk(data);
      const nullIndex = text.indexOf('\0');
      
      if (nullIndex > 0) {
        const keyword = text.substring(0, nullIndex).toLowerCase();
        const value = text.substring(nullIndex + 1);
        
        // 查找chara字段（角色数据）
        if (keyword === 'chara') {
          try {
            // 使用UTF-8安全的Base64解码（支持中文）
            const charaData = base64ToJson(value);
            return charaData;
          } catch (error) {
            console.error('Error parsing character data:', error);
            return null;
          }
        }
      }
    }
    
    if (type === 'IEND') break;
    i += 12 + length;
  }
  
  return null;
}

/**
 * UTF-8安全的Base64解码函数
 * 解决JavaScript原生atob()不支持中文的问题
 * @param {string} base64String - Base64编码的字符串
 * @returns {Object} - 解码后的JSON对象
 */
function base64ToJson(base64String) {
  // 步骤1: 使用atob解码为二进制字符串（Latin1编码）
  const binaryString = atob(base64String);
  
  // 步骤2: 转换为字节数组
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  
  // 步骤3: 使用TextDecoder正确解码UTF-8
  const decoder = new TextDecoder('utf-8');
  const utf8String = decoder.decode(bytes);
  
  // 步骤4: 解析JSON
  return JSON.parse(utf8String);
}

/**
 * UTF-8安全的Base64解码（返回字符串）
 * 用于exif_b64等包含中文的二进制数据
 * @param {string} base64String - Base64编码的字符串  
 * @returns {Uint8Array} - 解码后的字节数组
 */
function base64ToBytes(base64String) {
  const binaryString = atob(base64String);
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes;
}